File: apps/api/app/schemas/timeline.py
from datetime import date
from typing import Literal

from pydantic import BaseModel

# food 現在也有真的資料表了（food_records），跟 ai_scan/health_journal 一樣
# 併進時間軸；ai_scan/health_journal 不是每筆分析都自動算一筆時間軸事件——
# 只有使用者按過「加入時間軸」/「加入健康日誌」（added_to_timeline=True）
# 的才會被 timeline.py 撈進來，這點跟 vaccine/report/food 無條件全撈不一樣
TimelineItemType = Literal["food", "vaccine", "report", "ai_scan", "health_journal"]


# 時間軸上的一筆項目——把不同來源的資料表攤平成同一種格式，前端不用管背後是
# 哪張表，只要照 type 選 icon 就好
class TimelineItemOut(BaseModel):
    type: TimelineItemType
    id: int
    date: date
    # "HH:MM"，只有來源資料表真的存了時間的才有值（食物/AI 分析用的是
    # datetime 欄位）；疫苗/健檢/健康日誌只存日期，沒有實際時間可以顯示，
    # 這裡就是 None，前端要對這種情況防呆，不要瞎掰一個時間出來
    time: str | None = None
    title: str
    summary: str | None = None
    # 只有食物/AI 分析這種本來就有拍照的類型才會有縮圖
    image_url: str | None = None

File: apps/web/src/lib/timeline.ts
import {
  Camera,
  Smile,
  Stethoscope,
  Syringe,
  Utensils,
  type LucideIcon,
} from "lucide-react";
import { apiFetch } from "./api";

// 跟後端 app/schemas/timeline.py 的 TimelineItemOut 對齊——時間軸只是把
// food_records / vaccine_records / report_records / ai_scan_logs /
// health_journal_logs 這幾張表讀出來、依日期攤平合併，不是自己存了一份
// 新資料，之後有更多功能要進時間軸，type 才會跟著多起來
export type TimelineItemType =
  | "food"
  | "vaccine"
  | "report"
  | "ai_scan"
  | "health_journal";

export type TimelineItem = {
  type: TimelineItemType;
  id: number;
  date: string; // "YYYY-MM-DD"
  // "HH:MM"——只有食物/AI 分析這種來源資料表本來就存了時間的才有值，
  // 疫苗/健檢/健康日誌只存日期，這裡是 null，畫面要對這種情況防呆
  time: string | null;
  title: string;
  summary: string | null;
  image_url: string | null;
};

export function fetchTimeline(petId: number) {
  return apiFetch<TimelineItem[]>(`/timeline/${petId}`);
}

// 每種類型在時間軸卡片上的圖示/配色，照圖片的分類配色對照（食物橘、健檢綠、
// AI 分析紫、健康日誌黃、疫苗紅）
export const timelineTypeMeta: Record<
  TimelineItemType,
  { Icon: LucideIcon; iconClass: string; badgeClass: string }
> = {
  food: {
    Icon: Utensils,
    iconClass: "bg-[#fbe4cf] text-[#d9834f]",
    badgeClass: "bg-[#fbe4cf] text-[#d9834f]",
  },
  report: {
    Icon: Stethoscope,
    iconClass: "bg-[#dcf2ec] text-[#3fa88f]",
    badgeClass: "bg-[#dcf2ec] text-[#3fa88f]",
  },
  ai_scan: {
    Icon: Camera,
    iconClass: "bg-[#eae3f7] text-[#8462c9]",
    badgeClass: "bg-[#eae3f7] text-[#8462c9]",
  },
  health_journal: {
    Icon: Smile,
    iconClass: "bg-[#fdf0cf] text-[#c99a3f]",
    badgeClass: "bg-[#fdf0cf] text-[#c99a3f]",
  },
  vaccine: {
    Icon: Syringe,
    iconClass: "bg-[#fbe4de] text-[#c9503f]",
    badgeClass: "bg-[#fbe4de] text-[#c9503f]",
  },
};
